fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/DimensionsKeyExpression.java

```diff
@@ -0,0 +1,255 @@
+ */
+
+package com.apple.foundationdb.record.metadata.expressions;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.ObjectPlanHash;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.RecordMetaDataProto;
+import com.apple.foundationdb.record.metadata.Key;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecord;
+import com.apple.foundationdb.record.query.plan.cascades.KeyExpressionVisitor;
+import com.apple.foundationdb.record.util.HashUtils;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A key expression that divides into three parts that represent the constituent parts of a multidimensional index.
+ * Zero or more <em>prefix</em> columns determine a range within which an R-tree structure is maintained.
+ * The <em>dimensions</em> columns determine a set of expressions that span an multidimensional canvas. All
+ * such columns must be of a numeric type.
+ * The remaining (up to <code>Index.getColumnSize()</code>) <em>rest</em> are stored within the R-tree as key-rest or
+ * value.
+ */
+@API(API.Status.MAINTAINED)
+public class DimensionsKeyExpression extends BaseKeyExpression implements KeyExpressionWithChild {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Dimension-Key-Expression");
+
+    @Nonnull
+    private final KeyExpression wholeKey;
+    private final int prefixSize;
+    private final int dimensionsSize;
+
+
+    public DimensionsKeyExpression(@Nonnull final KeyExpression wholeKey,
+                                   final int prefixSize,
+                                   final int dimensionsSize) {
+        this.wholeKey = wholeKey;
+        this.prefixSize = prefixSize;
+        this.dimensionsSize = dimensionsSize;
+    }
+
+    public DimensionsKeyExpression(@Nonnull RecordMetaDataProto.Dimensions dimensions) throws DeserializationException {
+        this(KeyExpression.fromProto(dimensions.getWholeKey()), dimensions.getPrefixSize(), dimensions.getDimensionsSize());
+    }
+
+    public int getPrefixSize() {
+        return prefixSize;
+    }
+
+    public int getDimensionsSize() {
+        return dimensionsSize;
+    }
+
+    @Nonnull
+    @Override
+    public <M extends Message> List<Key.Evaluated> evaluateMessage(@Nullable final FDBRecord<M> record, @Nullable final Message message) {
+        return getWholeKey().evaluateMessage(record, message);
+    }
+
+    @Override
+    public boolean createsDuplicates() {
+        return getWholeKey().createsDuplicates();
+    }
+
+    @Override
+    public List<Descriptors.FieldDescriptor> validate(@Nonnull Descriptors.Descriptor descriptor) {
+        return getWholeKey().validate(descriptor);
```

Comment:
Some simple in-bounds sanity checking is possible on the two sizes.

Reply:
done.